pages/sitemap.xml.ts

```diff
@@ -1,56 +1,115 @@
-        ${paths
-        .map((p) => p.replace(/^pages\//, '')) // get rid of /pages
-        .map((p) => p.replace(/\.(tsx|ts)$/, '')) // get rid of file extension
-        .map((path) => `
-                <url>
-                    <loc>${`${url}/${path === 'index' ? '' : path}`}</loc>
-                </url>
-            `)
-        .join('')}
-        ${stories
-        .map(({ slug }) => `
-                <url>
-                    <loc>${`${url}/${slug}`}</loc>
-                </url>
-            `)
-        .join('')}
-    </urlset>
-`;
-
-class Sitemap extends React.Component {
-    static async getInitialProps(ctx: NextPageContext) {
-        const { res, req } = ctx;
-
-        if (!req || !res) { // client side
-            return null;
+import { Category, Story } from '@prezly/sdk/dist/types';
+import { NextPage, NextPageContext } from 'next';
+
+type Url = {
+    loc: string;
+    changefreq?: string;
+    priority?: string;
+};
+
+class SitemapBuilder {
+    private baseUrl: string;
+
+    private urls: Url[] = [];
+
+    constructor(baseUrl: string) {
+        this.baseUrl = baseUrl;
+    }
+
+    addUrl(loc: string) {
+        this.urls.push({
+            loc: this.baseUrl + loc,
+            changefreq: SitemapBuilder.guessFrequency(loc),
+            priority: SitemapBuilder.guessPriority(loc),
+        });
+    }
+
+    private static guessFrequency(loc: string) {
+        if (loc === '/') return 'daily';
+
+        return 'weekly';
+    }
+
+    private static guessPriority(loc: string) {
+        if (loc === '/') return '0.9';
+
+        if (loc.startsWith('/category/')) {
+            return '0.8';
         }
 
-        const paths = await globby([
-            'pages/**/*{.tsx,.ts}',
-            '!pages/**/[*', // ignore dynamic paths
-            '!pages/_*.js', // ignore nextjs special paths
-            '!pages/api', // ignore api routes
-        ]);
+        return '0.7';
+    }
 
-        const url = req.headers.host || '/';
+    static serializeLoc(url: Url) {
+        return [
+            '<url>',
+            url.loc && `\t<loc>${url.loc}</loc>`,
+            url.changefreq && `\t<changefreq>${url.changefreq}</changefreq>`,
+            url.priority && `\t<priority>${url.priority}</priority>`,
+            '</url>',
+        ]
+            .filter(Boolean)
+            .join('\n');
+    }
 
-        const api = getPrezlyApi(req);
-        const stories = await api.getAllStoriesNoLimit();
+    serialize() {
+        return [
+            '<?xml version="1.0" encoding="UTF-8" ?>',
+            '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:news="http://www.google.com/schemas/sitemap-news/0.9" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">',
+            this.urls.map(SitemapBuilder.serializeLoc).join('\n'),
+            '</urlset>',
+        ].join('\n');
+    }
+}
+
+const createPaths = (
+    stories: Array<Story>,
+    categories: Array<Category>,
+) => {
```

Comment:
Eventually we might wanna add something like this to the API, cause it'd be easier for people that want to use SSG

Reply:
That's true 👍 Api or at least to sdk/sdk wrapper. As we currently have in website-nextjs services.

Reply:
yeah I always call the wrapper "api"